src/gpu/structs.rs

```diff
@@ -0,0 +1,32 @@
+use paired::{CurveAffine, CurveProjective};
+use ff::{PrimeField};
+use ocl::traits::OclPrm;
+
+// Everything that needs to be copied to GPU needs to implement OclPrm trait.
+// This module implements a generic OclPrm version of PrimeField, CurveAffine and CurveProjective
+// so that they can be copied to GPU.
+// A transmute from [PrimeField] to [PrimeFieldStruct] is safe as their sizes are equal.
+// Also safe for [CurveAffine] to [CurveAffineStruct] and [CurveProjective] to [CurveProjectiveStruct]
+// It costs less memory to just transmute the data sources for both FFT and Multiexp to their
+// OpenCL friendly equivalents instead of mapping them to a new array with each element casted.
+
+#[derive(PartialEq, Debug, Clone, Copy)]
+pub struct PrimeFieldStruct<T>(pub T);
```

Comment:
I think it would be safer to use `#[repr(transparent)]` https://github.com/rust-lang/rfcs/blob/master/text/1758-repr-transparent.md for these, as I don't think the gurantees are quite a stable for transmutes of these.